module/bdev/lvol/vbdev_lvol.c

```diff
@@ -2087,4 +2089,182 @@ vbdev_lvol_shallow_copy(struct spdk_lvol *lvol, const char *bdev_name,
+	num_clusters = spdk_divide_round_up(next_offset - req->current_offset, req->fragmap.cluster_size);
+
+	for (uint64_t i = 0; i < num_clusters; i++) {
+		spdk_bit_array_set(req->fragmap.map, start_cluster + i);
+	}
+	req->fragmap.num_allocated_clusters += num_clusters;
+
+	req->current_offset = next_offset;
+
+	if (req->current_offset == req->offset + req->size) {
+		get_fragmap_done(req, 0, NULL);
+		return;
+	}
+
+	rc = spdk_bdev_seek_data(req->bdev_desc, req->bdev_io_channel,
+				 spdk_divide_round_up(req->current_offset, req->fragmap.block_size),
+				 seek_data_done_cb, req);
+	if (rc != 0) {
+		get_fragmap_done(req, rc, "failed to seek data");
+	}
+}
+
+static void
+dummy_bdev_event_cb(enum spdk_bdev_event_type type, struct spdk_bdev *bdev, void *ctx)
+{
+}
+
+void
+vbdev_lvol_get_fragmap(struct spdk_lvol *lvol, uint64_t offset, uint64_t size,
+		       spdk_lvol_op_with_fragmap_handle_complete cb_fn, void *cb_arg)
+{
+	struct spdk_bdev_desc *desc;
+	struct spdk_io_channel *channel;
+	struct spdk_bit_array *fragmap;
+	struct spdk_fragmap_req *req;
+	uint64_t cluster_size, num_clusters, block_size, num_blocks, lvol_size, segment_size;
+	int rc;
+
+	// Create a bitmap recording the allocated clusters
+	cluster_size = spdk_bs_get_cluster_size(lvol->lvol_store->blobstore);
+	block_size = spdk_bdev_get_block_size(lvol->bdev);
+	num_blocks = spdk_bdev_get_num_blocks(lvol->bdev);
+	lvol_size = num_blocks * block_size;
+
+	if (offset + size > lvol_size) {
+		SPDK_ERRLOG("offset %lu and size %lu exceed lvol size %lu\n",
+			    offset, size, lvol_size);
+		cb_fn(cb_arg, NULL, -EINVAL);
+		return;
+	}
+
+	segment_size = size;
+	if (size == 0) {
+		segment_size = lvol_size;
+	}
+
+	if (!spdk_is_divisible_by(offset, cluster_size) ||
+	    !spdk_is_divisible_by(segment_size, cluster_size)) {
+		SPDK_ERRLOG("offset %lu and size %lu must be a multiple of cluster size %lu\n",
+			    offset, segment_size, cluster_size);
+		cb_fn(cb_arg, NULL, -EINVAL);
+		return;
+	}
+
+	num_clusters = spdk_divide_round_up(segment_size, cluster_size);
+	fragmap = spdk_bit_array_create(num_clusters);
+	if (fragmap == NULL) {
+		SPDK_ERRLOG("failed to allocate fragmap with num_clusters %lu\n", num_clusters);
+		cb_fn(cb_arg, NULL, -ENOMEM);
+		return;
+	}
+
+	// Construct a fragmap of the lvol
+	rc = spdk_bdev_open_ext(lvol->bdev->name, false,
+				dummy_bdev_event_cb, NULL, &desc);
+	if (rc != 0) {
+		SPDK_ERRLOG("could not open bdev %s\n", lvol->bdev->name);
+		cb_fn(cb_arg, NULL, rc);
+		return;
+	}
+
+	channel = spdk_bdev_get_io_channel(desc);
+	if (channel == NULL) {
+		spdk_bdev_close(desc);
+		SPDK_ERRLOG("could not allocate I/O channel.\n");
+		cb_fn(cb_arg, NULL, -ENOMEM);
+		return;
+	}
+
+	req = spdk_zmalloc(sizeof(struct spdk_fragmap_req), 0, NULL,
+			  SPDK_ENV_LCORE_ID_ANY, SPDK_MALLOC_DMA);
+	if (req == NULL) {
+		SPDK_ERRLOG("could not allocate fragmap_io\n");
+		spdk_bit_array_free(&req->fragmap.map);
```

Comment:
you must free `fragmap`, `req->fragmap.map` has still not be assigned

Reply:
No need to free `fragmap`.
```
struct spdk_fragmap_req {
	struct spdk_bdev *bdev;
	struct spdk_bdev_desc *bdev_desc;
	struct spdk_io_channel *bdev_io_channel;
	//struct spdk_jsonrpc_request *request;

	struct spdk_fragmap fragmap;

	uint64_t offset;
	uint64_t size;
	uint64_t current_offset;

	spdk_lvol_op_with_fragmap_handle_complete cb_fn;
	void *cb_arg;
};
```

Reply:
Fixed. Thank you.